pkg/kv/kvserver/rangefeed/registry.go

```diff
@@ -303,6 +341,23 @@ func (r *registration) runOutputLoop(ctx context.Context, _forStacks roachpb.Ran
 	r.disconnect(roachpb.NewError(err))
 }
 
+// drainAllocations should be done after registration is disconnected from
+// processor to release all memory budget that its pending events hold.
+func (r *registration) drainAllocations(ctx context.Context) {
+	for {
+		select {
+		case e, ok := <-r.buf:
```

Comment:
I wonder if we'd want to close `r.buf` when the registration is done, to make sure there's no race condition where another event gets into the channel after we've drained it. If so, this simplifies to a `for e := range r.buf` loop.

Reply:
I tried this at some point but had issues with accessing closed channel in tests.

Reply:
Ok. Not important, as long as we're confident that we don't have any leaks.

Maybe we could consider adding some assertions that all allocation have been freed when we close the budget, or something. If they're expensive/dangerous we could consider only enabling then in race builds or something. We don't have to do this now though, I'm fine with adding it during stability.